### PR TITLE
add tests

### PR DESCRIPTION
**Pull request by nvergel**

### CODE REVIEW

File: portfolio/src/test/java/com/google/sps/servlets/MessageTest.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  public void testLessThanAndGreaterThan() {
+    Message message1 = new Message("0", "a", 1);
+    Message message2 = new Message("1", "b", 2);
+
+    Assert.assertTrue(message1.compareTo(message2) > 0);
+    Assert.assertTrue(message2.compareTo(message1) < 0);
+  }
+
+@Test
+  public void testEqual() {
```
Comment (bengaskagoogle) [inline portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:44]: Might be worth adding an actual 

boolean equals(message)

function rather than using compareTo as an equivalent class. Is there a natural ordering to messages? You already test compareTo above with a natural ordering. So define an equals() function and move this == 0 case to the test above.

File: portfolio/src/test/java/com/google/sps/servlets/MessageTest.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+package com.google.sps.servlets;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class MessageTest {
```
Comment (bengaskagoogle) [inline portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:23]: I went ahead and looked at the Message class as well. I'm curious what other functionality you think that you could add and test with this class? This is treated as a struct basically, with a compareTo defined to only rely on time created.

Is there any other useful functionality we can add to this Message class involving the other fields?

Comment (nvergel) [reply portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:23]: I put the messages in an ArrayList and to use Sort() I had to implement compareTo. I can't think of any other functionality.

File: portfolio/src/test/java/com/google/sps/servlets/MessageTest.java (hunk 1 above)
Comment (nvergel) [reply portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:44]: I just wanted to test compareTo, maybe the name of the test was confusing.

Comment (bengaskagoogle) [reply portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:44]: A relatively common thing is to explicitly name the method or endpoint being tested in the name of the test, along with a brief description of what exactly is being tested.

Comment (bengaskagoogle) [reply portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:44]: Also, why don't you go ahead and implement equals. Should be simple.

File: portfolio/src/test/java/com/google/sps/servlets/MessageTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2019 Google LLC
```
Comment (bengaskagoogle) [inline portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:1]: Remind me to reach out and confirm on this. Should this be 2020? It might matter depending on what things Google intends with this.

File: portfolio/src/test/java/com/google/sps/servlets/MessageTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    Message message3 = new Message("2", "c", 1);
+
+    Assert.assertTrue(message1.compareTo(message2) > 0);
+    Assert.assertTrue(message2.compareTo(message1) < 0);
+    Assert.assertEquals(message1.compareTo(message3), 0);
+    Assert.assertEquals(message3.compareTo(message1), 0);
+  }
+
+@Test
+  public void testEqual() {
```
Comment (bengaskagoogle) [inline portfolio/src/test/java/com/google/sps/servlets/MessageTest.java:47]: It feels a bit much considering the simplicity of the code, but normally we'd test various edge cases and extremes as well. 

So, empty string is an obvious edge case. Things like that. Kind of not super meaningful in the context but wanted to mention that.

Also, if you're bored look up Fuzz testing. Have you ever used these techniques or seen them? They are interesting.

No real actionable comment here. Just thinking. It's late at night, and I'm prone to rambling.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -15,9 +15,103 @@
...
+import java.util.Arrays;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> timesForMeeting = new ArrayList<TimeRange>();
+
+    // Meeting is too long  
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: So, an actual issue.

This is a pretty big function. And it is very, very good programming practice to break your code up into helper functions. These will be private helpers. A good rule of thumb is don't have any function longer than ~20 lines. It's not an absolute, but you should have a hard reason to break it.

Each of these functions should be private in this case, and the name should be a overview of what the logic actually does. I will insist on this.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -15,9 +15,103 @@
...
+    // Keep track of time slots for optional attendees
+    int beginSlotOptional = beginSlot;
+
+    for (Event event : events) {
+      // Check for conflicts
+      boolean noConflict = checkForConflict(event, request.getAttendees());
+      boolean noOptionalConflict = checkForConflict(event, request.getOptionalAttendees());
+
+      // If optionals can be included do original search method
+      if (optionalsIncluded) {
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: In some of this your reaching a fairly deep level of nesting. You should look over this and consider how you can logically refactor flow to avoid this deep nesting.

Comment (bengaskagoogle) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: Might be worth drawing a truth table, btw. That's fairly useful for complex conditionals such as this.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -15,9 +15,103 @@
...
+  }
+
+  /* Checks if the set of an event's attendees does not intersect the
+   * set of the request's attendees
+   */
+  public boolean checkForConflict(Event event, Collection<String> attendees) {
+    return Collections.disjoint(event.getAttendees(), attendees);
+  }
+
+  public int addSlotIfItFits(Event event, MeetingRequest request,
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:94]: Should this and checkForConflict be public? If it's not part of your API, make it private. If it is, add a comment to addSlotIfItFits.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -15,9 +15,103 @@
...
+    if ( slotFits(event, request, beginSlot) ) {
+      timesForMeeting.add(TimeRange.fromStartEnd(beginSlot, endSlot, false));
+    }
+
+    // Start new slot after event ends
+    if (beginSlot < event.getWhen().end()) {
+      beginSlot = event.getWhen().end();
+    }
+
+    return beginSlot;
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:109]: I don't like that this makes it so that if an item is not added, the result of that fact is implicit. 

This function either does nothing, silently. Also, in your logic beginSlot becomes quite complicated over the various sets of functions it is passed to.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -15,9 +15,103 @@
...
+  }
+
+  // Returns false if the current slot is too short for the request
+  public boolean slotFits(Event event, MeetingRequest request, int beginSlot) {
+    int endSlot = event.getWhen().start();
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:114]: I would recommend moving this into a variable that called "eventLength" or something, and doing the math up here for that. Than your result length is "eventLength >= duration" which is clearer than what you have now.

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 10]
```diff
@@ -270,5 +271,130 @@ public void notEnoughRoom() {
...
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void everyOptionalIsConsidered() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|     |--B--|
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:354]: Very fancy!

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 11]
```diff
@@ -270,5 +271,130 @@ public void notEnoughRoom() {
...
+    Collection<TimeRange> actual = query.query(events, request);
+    Collection<TimeRange> expected =
+        Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+            TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+            TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  /* I think this test violates the assumption of what should be returned
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:378]: What's going on here?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -15,9 +15,103 @@
...
+        if ( noConflict && noOptionalConflict) {
+          continue;
+        }
+
+        beginSlot = addSlotIfItFits(event, request, beginSlot, timesForMeeting);
+
+      } else {
+        // Check to see if optional attendees can be included
+        if (noOptionalConflict) {
+          if (slotFits(event, request, beginSlotOptional)) {
```
Comment (bengaskagoogle) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:58]: And things like this can be simplified. Having a guard that only assigns a boolean can get rid of the guard to make things clearer.

**Review by bengaskagoogle — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:47]: I figured out how to improve the logic and substantially reduce the nesting. I think I just needed a fresh go at this problem.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:29]: Fixed

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: fixed

Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: If it is private it does not need comments?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: this is void now

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:94]: done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Outdated -> New implementation removed nesting

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 11 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:378]: To quote the tutorial,
"The basic functionality of optional attendees is that if one or more time slots exists so that both mandatory and optional attendees can attend, return those time slots. Otherwise, return the time slots that fit just the mandatory attendees."

In this test the request only has optional attendees who do not have any free time. Therefore based on what query should do, it will return the whole day as a time slot because there are no mandatory attendees keeping it from not being scheduled. Instead the test checks that we return empty time slot because of the optional attendees. I think this violates the expected behavior.

Comment (bengaskagoogle) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:378]: Don't comment this out though. This means it is something to clarify, not ignore.

Comment (bengaskagoogle) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:378]: Just to clarify, was this written in regards to this test:

`No mandatory attendees, just two optional attendees with no gaps in their schedules. query should return that no time is available.`

Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:378]: Yeah that is what was written

Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:378]: I could add an if statement that checks if mandatory attendees is empty to pass this test but I just felt that it would violate the assumed behavior

Comment (bengaskagoogle) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:378]: I think your test should not have any mandatory attendees than. It seems like you pass them into the event at creation time. PERSON_A and PERSON_B, right?

**Review by bengaskagoogle — APPROVED** [review]

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 11 above)
Comment (nvergel) [reply walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: `request.addOptionalAttendee(PERSON_A);` adds them as optional attendees
